src/Microsoft.Data/StringColumn.cs

```diff
@@ -0,0 +1,199 @@
+        {
+            List<int[]> bufferSortIndices = new List<int[]>(_stringBuffers.Count);
+            foreach (var buffer in _stringBuffers)
+            {
+                var sortIndices = new int[buffer.Count];
+                for (int i = 0; i < buffer.Count; i++) sortIndices[i] = i;
+                // TODO: Refactor the sort routine to also work with IList?
+                string[] array = buffer.ToArray();
+                IntrospectiveSort(array, array.Length, sortIndices, comparer);
+                bufferSortIndices.Add(sortIndices);
+            }
+            // Simple merge sort to build the full column's sort indices
+            SortedDictionary<string, List<Tuple<int, int>>> heapOfValueAndListOfTupleOfSortAndBufferIndex = new SortedDictionary<string, List<Tuple<int, int>>>(comparer);
+            var buffers = _stringBuffers;
+            for (int i = 0; i < buffers.Count; i++)
+            {
+                List<string> buffer = buffers[i];
+                string value = buffer[bufferSortIndices[i][0]];
+                if (heapOfValueAndListOfTupleOfSortAndBufferIndex.ContainsKey(value))
+                {
+                    heapOfValueAndListOfTupleOfSortAndBufferIndex[value].Add(new Tuple<int, int>(0, i));
+                }
+                else
+                {
+                    heapOfValueAndListOfTupleOfSortAndBufferIndex.Add(value, new List<Tuple<int, int>>() {new Tuple<int, int>(0, i) });
+                }
+            }
+            columnSortIndices = new PrimitiveColumn<long>("SortIndices");
+            GetBufferSortIndex getBufferSortIndex = new GetBufferSortIndex((int bufferIndex, int sortIndex) => bufferSortIndices[bufferIndex][sortIndex]);
+            GetValueAtBuffer<string> getValueAtBuffer = new GetValueAtBuffer<string>((int bufferIndex, int sortIndex) => _stringBuffers[bufferIndex][bufferSortIndices[bufferIndex][sortIndex]]);
+            GetBufferLengthAtIndex getBufferLengthAtIndex = new GetBufferLengthAtIndex((int bufferIndex) => bufferSortIndices[bufferIndex].Length);
+            PopulateColumnSortIndicesWithHeap(heapOfValueAndListOfTupleOfSortAndBufferIndex, columnSortIndices, getBufferSortIndex, getValueAtBuffer, getBufferLengthAtIndex);
+        }
+
+        public override BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false)
+        {
+            if (!(mapIndices is null))
+            {
+                if (mapIndices.DataType != typeof(long)) throw new ArgumentException($"Expected sortIndices to be a PrimitiveColumn<long>");
+                if (mapIndices.Length != Length) throw new ArgumentException(strings.MismatchedColumnLengths, nameof(mapIndices));
+                return _Clone(mapIndices as PrimitiveColumn<long>, invertMapIndices);
+            }
+            return _Clone();
+        }
+
+        public StringColumn _Clone(PrimitiveColumn<long> mapIndices = null, bool invertMapIndex = false)
```

Comment:
Why does this public method start with an underscore? We don't use underscores in public member names.

Reply:
Oversight while refactoring. Fixed